grammar/rainerscript.c

```diff
@@ -2377,9 +2378,9 @@ doFunct_Lookup(struct cnffunc *__restrict__ const func,
 		return;
 	}
 	cnfexprEval(func->expr[1], &srcVal, usrptr, pWti);
-	pthread_rwlock_rdlock(&((lookup_ref_t*)func->funcdata)->rwlock);
-	pthread_rwlock_unlock(&((lookup_ref_t*)func->funcdata)->rwlock);
-	lookup_table = ((lookup_ref_t*)func->funcdata)->self;
+	lookup_table_ref = (lookup_ref_t*) func->funcdata;
+	pthread_rwlock_rdlock(&lookup_table_ref->rwlock);
+	lookup_table = lookup_table_ref->self;
```

Comment:
Q, now that I dig somewhat deeper - If you don't know out of your head, I can look it up.

Is it safe to use `lookup_table` if there was a reload before use (not guarded). In other words: is the old table freed after reload?

Reply:
Yep, we free old-table after reuse (lookup.c:576).

This was a pretty bad race.
All usage sites need to hold shared-lock to prevent concurrent-free race.

Reply:
do you remember that old strange issue https://github.com/rsyslog/rsyslog/issues/1071? Might this fix here be related? Doesn't sound so, but I think in theory it could.